[Cache][FrameworkBundle] add `cache:pool:invalidate-tags` command

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #40218
| License       | MIT
| Doc PR        | symfony/symfony-docs#16328

Adds a `cache:pool:invalidate-tags` command. 

## Usage

```bash
bin/console cache:pool:invalidate-tags tag1 # invalidates tag1 from all taggable pools
bin/console cache:pool:invalidate-tags tag1 tag2 # invalidates tag1 & tag2 from all taggable pools
bin/console cache:pool:invalidate-tags tag1 tag2 --pool=cache.app # invalidates tag1 & tag2 from cache.app pool
bin/console cache:pool:invalidate-tags tag1 tag2 -p cache1 -p cache2 # invalidates tag1 & tag2 from cache1 & cache2 pools
```

## TODO

- [x] tests
- [x] account for #44682 (once merged up to 6.1)

kbond committed:
[Cache][FrameworkBundle] add `cache:pool:invalidate-tags` command

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 ---
 
  * Environment variable `SYMFONY_IDE` is read by default when `framework.ide` config is not set.
+ * Add `cache:pool:invalidate-tags` command.
```
no final dot please (you can remove it from the previous line also)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolInvalidateTagsCommand.php:
```diff
@@ -0,0 +1,98 @@
+            return $this->pools;
+        }
+
+        foreach ($this->pools as $key => $pool) {
```
instead of iterating over all pools here, what about injection a ServiceProviderInterface to the constructor?

kbond replied:
I made this change. What do you think about dropping the pass entirely and just using `tagged_locator('cache.taggable', 'pool')` for the command's argument? (We would lose the ability to ensure the pools are `TagAwareCacheInterface`'s at compile time)

nicolas-grekas replied:
good idea, we should not ensure pools are TagAwareCacheInterface, that's not what DI config is for

nicolas-grekas commented on src/Symfony/Component/Cache/DependencyInjection/CachePoolTaggablePass.php:
```diff
@@ -0,0 +1,71 @@
+
+                $class = $definition->getClass();
+
+                if (!$reflection = $container->getReflectionClass($class)) {
```
no need for the reflection checks, let's save loading classes at compile time

nicolas-grekas commented:
Please implement the `complete()` for pools :)

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolInvalidateTagsCommand.php:
```diff
@@ -0,0 +1,85 @@
+            try {
+                $pool = $this->pools->get($name);
+            } catch (ServiceNotFoundException $e) {
+                throw new \InvalidArgumentException(sprintf('Pool "%s" not found or is not taggable.', $name), previous: $e);
+            }
+
+            $io->comment(sprintf('Invalidating tag(s): <info>%s</info> from pool <comment>%s</comment>.', $tagList, $name));
+
+            if (!$pool->invalidateTags($tags)) {
+                throw new \Exception(sprintf('Cache tag(s) "%s" could not be invalidated for pool "%s".', $tagList, $name));
```
shouldnt we try to clear as many pools as possible?

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolInvalidateTagsCommand.php:
```diff
@@ -0,0 +1,85 @@
+    {
+        $this
+            ->addArgument('tags', InputArgument::IS_ARRAY | InputArgument::REQUIRED, 'The tags to invalidate')
+            ->addOption('pool', 'p', InputOption::VALUE_REQUIRED, 'The pool to invalidate on')
```
VALUE_IS_ARRAY also?

kbond commented:
Thanks for the reviews! I made the following changes:
1. Implemented `complete()`
2. Replaced compiler pass with `tagged_locator`
3. Made `--pool` option an array
4. Exceptions thrown when invalidating tags are displayed but don't cause the command to fail until all pools are attempted.

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolInvalidateTagsCommand.php:
```diff
@@ -0,0 +1,118 @@
+            $io->comment(sprintf('Invalidating tag(s): <info>%s</info> from pool <comment>%s</comment>.', $tagList, $name));
+
+            try {
+                $this->invalidateTags($name, $tags);
```
inline it, and work directly with $io? i'd only catch `$pool->invalidateTags($tags)`, the others are useless backtraces.

kbond replied:
So for an invalid/missing pool, throw the exception directly and don't continue?

ro0NL replied:
i mean, instead of throwing in invalidateTags, then catching here, to then print using $io ..., it could print directly using $io.

In that sense you can pass $io to invalidateTags as well (but it's a single used method, hence the suggestion to inline it).

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Command/CachePoolInvalidateTagsCommand.php:
```diff
@@ -0,0 +1,118 @@
+            try {
+                $this->invalidateTags($name, $tags);
+            } catch (\Throwable $e) {
+                $this->getApplication() ? $this->getApplication()->renderThrowable($e, $output) : $io->error($e->getMessage());
```
meh :) would be nice if $Io->errror can accept string|\Throwable

kbond commented:
Tests added. I believe this PR to be ready.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/Command/CachePoolInvalidateTagsCommandTest.php:
```diff
@@ -0,0 +1,138 @@
+use Symfony\Component\DependencyInjection\ServiceLocator;
+use Symfony\Contracts\Cache\TagAwareCacheInterface;
+
+final class CachePoolInvalidateTagsCommandTest extends TestCase
```
we don't make tests `final` as that serves no purpose

94noni commented:
nice addition :)
May I ask you if https://github.com/symfony/symfony/pull/37880 could be reconsidered for v6? thank you

chalasr commented:
@kbond can you rebase this one?

kbond commented:
Rebased.

chalasr commented:
Thank you @kbond.
